jetty-websocket/websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/server/ServerCloseTest.java

```diff
@@ -275,4 +276,27 @@ public void testOpenSessionCleanup() throws Exception
             close(session);
         }
     }
+
+    @Test
+    public void testHardClose() throws Exception
+    {
+        // Testing WebSocketSession.close() in onClosed() does not cause deadlock.
+        ClientUpgradeRequest request = new ClientUpgradeRequest();
+        request.setSubProtocols("closeInOnClose");
+        CloseTrackingEndpoint clientEndpoint = new CloseTrackingEndpoint();
+
+        URI wsUri = WSURI.toWebsocket(server.getURI().resolve("/ws"));
+        client.connect(clientEndpoint, wsUri, request).get(5, SECONDS);
+
+        // Hard close from the server. Server onClosed() will try to close again which should be a NOOP.
+        AbstractCloseEndpoint serverEndpoint = serverEndpointCreator.pollLastCreated();
+        serverEndpoint.getSession().close(StatusCode.SHUTDOWN, "SHUTDOWN hard close");
+
+        // Verify that client got close
+        clientEndpoint.assertReceivedCloseEvent(5000, is(StatusCode.SHUTDOWN), containsString("SHUTDOWN hard close"));
+
+        // Verify that server socket got close event
+        assertTrue(serverEndpoint.closeLatch.await(5, SECONDS));
+        assertThat(serverEndpoint.closeStatusCode, is(StatusCode.SHUTDOWN));
+    }
```

Comment:
Please add a test where 2 closes are performed from 2 different threads. The path I would like to be tested is:

Thread T1 calls `close()` triggering abnormal close so that `onWebSocketClose()` gets called.
In `onWebSocketClose()` spawn a new thread T2 that calls close() and wait for T2 to finish.
Expect T1 to finish (was deadlocking before).

Reply:
added another test for this below here `testSecondCloseFromOnClosedInNewThread`